source/language/openpulse.rst

```diff
@@ -0,0 +1,316 @@
+format:
+
+- It is more readable
+- Absolute time is handled through built-in timekeeping instructions
+- Gates and classical instructions can be mixed in with the pulses to create far richer calibrations
+- Pulse definitions are tied to circuit instructions rather than circuit programs
+- Richer handling of frames
+
+Channels
+--------
+
+Channels map to a hardware resource, which can play pulses to manipulate a qubit
+or capture signal from the qubit to perform a measurement. There is a many-to-many
+relationship between qubits and channels. One qubit may have multiple channels
+connecting to it. Pulses on different channels would have different physical
+interactions with that qubit. A channel may also have many qubits. For instance,
+a channel could manipulate the coupling between two neighboring qubits, or
+could even reference multiple qubits in a chain.
+
+Channels are vendor-specific and device-specific. It is expected that vendors
+of quantum hardware provide the appropriate channel names and qubit mappings
+for use by end users. Each channel may also have associated static settings,
+such as local-oscillator frequencies, which do not vary throughout program
+execution. Again it is expected that vendors of quantum hardware provide a
+method for manipulating those static settings if appropriate.
+
+There are two kinds of channels: transmit channels (sending input to a quantum
+device) and receive channels (reading output from a quantum device).
+
+The ``channel`` type is a compile-time type representing a channel.
+
+A value of type ``channel`` is retreived with the special functions
+``transmit`` and ``receive``. These functions take an optional string argument
+followed by 1 or more physical qubits.
+
+The hardware vendor, during compilation, will substitute valid calls to
+``transmit`` and ``receive`` with their corresponding channels.
+
+A recommended convention: calling ``transmit`` with a single qubit argument
+should return the channel for driving single qubit gates, and calling
+``receive`` with a single qubit argument should return the channel for reading
+out the state of that qubit.
+
+.. code-block:: none
+
+   // Get the default input channel for qubit 0
+   transmit(%0)
+   // Get the default output channel for qubit 1
+   receive(%1)
+   // Get an input channel named "measure" for qubit 1
+   // Could be used to specify the measurement stimulus channel
+   transmit("measure", %1)
+   // Get an output channel named "cr1" for qubit 0
+   receive("cr1", %0)
+   // Get the default output channel for qubits 0 and 1
+   receive(%0, %1)
+
+Frames
+------
+
+It turns out to be quite useful to keep track of a set of carrier signals
+throughout the execution of a program. These carrier signals are called "frames"
+and are defined by :math:$A*e^{i\left(2\pi f t + \theta\right)}$, where `f` is
+frequency, `theta` is phase, and `A` is a scaling factor.
+
+Frames are attached to channels by a 1:1 correspondence. Every frame must 
```

Comment:
Is it 1:1? Clearly a channel can have multiple frames, but a frame can only have one channel.

Reply:
The `frameof` operation only really works if there is a 1:1 mapping between channels and frames. I suppose nothing prevents us from creating additional frames that are not attached to channels.

Reply:
It is perhaps worth noting that the existing OpenPulse spec does not allow for multiple frames per channel, but I find this to be problematic and worth changing.

Reply:
I agree frame-per-channel was a mistake and necessitates having every channel be a "logical" overlay channel of which many would be mapped to the same hardware channel to support multiple frames. This is in turn is very inflexible and makes applying a pulse at the frame of another channel very difficult.

Solutions I see:
 - Frames are explicit values that can be manipulated/assigned to channels. This is the approach I took in the document circulated privately.
 - Channels are derivable from existing channels with new frame relationships. This leads to an explosion of "channels" and the line between hardware/software entities may become blurred.

Reply:
*edit* - I now see the ability to supply a frame below to `play`.

Reply:
I'm struggling a bit with why we need both channels and frames. The notion of channels that "map to hardware resources" seems to mix in implementation concerns and makes it feel like the channel is either a data converter or a sequencer.  Can't we get away with claiming that everything is just played on an abstract frame at this point?  For example, if a hardware vendor supplies multiple frames that map to the same data converter then it's up to them to support that with multiple sequencers and an adder tree, or to merge the waveforms in software or however they want to support it. Or is the notion that you have a more limited set of frames and cover the combinatorial crosstalk issue with the product of a channel and frame?

Reply:
I am hardly tied to the concept of a frame. The problems that channels + frames are attempting to solve are scheduling constraints in hardware. 

If I have multiple frames per-physical channel at the openpulse level it is not clear if resources are shared at scheduling time (this is a philosophical choice if we wish to make the language-aware of this, internally we have a system like this and we refer to these as logical channels). I am ok with this as a lowering step in the compiler could be to map multiple "frames" to a single frame via sideband mixing (or error if constraints cannot be met) in software as you describe.

However, it is not possible to express hardware that is capable of supporting multiple togglable frames going to a single output DAC and this responsibility would have to fall to a further lower-level representation to express toggling frames on a single channel. 

In either representation (frames, or frames+channels) it is important that the frame exist as a persistent entity that accumulates states throughout the execution in real-time to enable mapping of input arguments to carrier phases. It is also necessary to be able to reference and manipulate the state of these carrier frames across channels. I believe both proposals contain this (eg., `playwfr`, `SWAP-PHASE`).